File: r/hamming/test_hamming.R
source("./hamming.R")
library(testthat)

context("hamming")

test_that("identical strands", {
  strand1 <- "A"
  strand2 <- "A"
  expect_equal(hamming(strand1, strand2), 0)
})

test_that("long identical strands", {
  strand1 <- "GGACTGA"
  strand2 <- "GGACTGA"
  expect_equal(hamming(strand1, strand2), 0)
})

test_that("complete distance in single nucleotide strands", {
  strand1 <- "A"
  strand2 <- "G"
  expect_equal(hamming(strand1, strand2), 1)
})

test_that("complete distance in small strands", {
  strand1 <- "AG"
  strand2 <- "CT"
  expect_equal(hamming(strand1, strand2), 2)
})

test_that("small distance in small strands", {
  strand1 <- "AT"
  strand2 <- "CT"
  expect_equal(hamming(strand1, strand2), 1)
})

test_that("small distance", {
  strand1 <- "GGACG"
  strand2 <- "GGTCG"
  expect_equal(hamming(strand1, strand2), 1)
})

test_that("small distance in long strands", {
  strand1 <- "ACCAGGG"
  strand2 <- "ACTATGG"
  expect_equal(hamming(strand1, strand2), 2)
})

test_that("non-unique character in first strand", {
  strand1 <- "AGA"
  strand2 <- "AGG"
  expect_equal(hamming(strand1, strand2), 1)
})

test_that("non-unique character in second strand", {
  strand1 <- "AGG"
  strand2 <- "AGA"
  expect_equal(hamming(strand1, strand2), 1)
})

test_that("same nucleotides in different positions", {
  strand1 <- "TAG"
  strand2 <- "GAT"
  expect_equal(hamming(strand1, strand2), 2)
})

test_that("large distance", {
  strand1 <- "GATACA"
  strand2 <- "GCATAA"
  expect_equal(hamming(strand1, strand2), 4)
})

test_that("empty strands", {
  strand1 <- ""
  strand2 <- ""
  expect_equal(hamming(strand1, strand2), 0)
})

test_that("disallow first strand longer", {
  strand1 <- "AATG"
  strand2 <- "AAA"
  expect_error(hamming(strand1, strand2))
})

test_that("disallow second strand longer", {
  strand1 <- "ATA"
  strand2 <- "AGTG"
  expect_error(hamming(strand1, strand2))
})

message("All tests passed for exercise: hamming")

File: zig/hamming/test_hamming.zig
const std = @import("std");
const testing = std.testing;

const hamming = @import("hamming.zig");
const DnaError = hamming.DnaError;

test "empty strands" {
    const expected = DnaError.EmptyDnaStrands;
    const actual = comptime hamming.compute("", "");
    try testing.expectError(expected, actual);
}

test "single letter identical strands" {
    const expected = 0;
    const actual = comptime try hamming.compute("A", "A");
    try testing.expectEqual(expected, actual);
}

test "single letter different strands" {
    const expected = 1;
    const actual = comptime try hamming.compute("G", "T");
    try testing.expectEqual(expected, actual);
}

test "long identical strands" {
    const expected = 0;
    const actual = comptime try hamming
        .compute("GGACTGAAATCTG", "GGACTGAAATCTG");
    try testing.expectEqual(expected, actual);
}

test "long different strands" {
    const expected = 9;
    const actual = comptime try hamming
        .compute("GGACGGATTCTG", "AGGACGGATTCT");
    try testing.expectEqual(expected, actual);
}

test "disallow first strand longer" {
    const expected = DnaError.UnequalDnaStrands;
    const actual = comptime hamming.compute("AATG", "AAA");
    try testing.expectError(expected, actual);
}

test "disallow second strand longer" {
    const expected = DnaError.UnequalDnaStrands;
    const actual = comptime hamming.compute("ATA", "AGTG");
    try testing.expectError(expected, actual);
}

test "disallow left empty strand" {
    const expected = DnaError.EmptyDnaStrands;
    const actual = comptime hamming.compute("", "G");
    try testing.expectError(expected, actual);
}

test "disallow right empty strand" {
    const expected = DnaError.EmptyDnaStrands;
    const actual = comptime hamming.compute("G", "");
    try testing.expectError(expected, actual);
}
